src/passes/I64ToI32Lowering.cpp

```diff
@@ -516,13 +516,22 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
 
   void lowerEqZInt64(Unary* curr) {
     TempVar highBits = fetchOutParam(curr->value);
-    replaceCurrent(
-      builder->makeBinary(
-        AndInt32,
-        builder->makeUnary(EqZInt32, builder->makeGetLocal(highBits, i32)),
-        builder->makeUnary(EqZInt32, curr->value)
+    TempVar lowBits = getTemp();
+    SetLocal* setLow = builder->makeSetLocal(lowBits, curr->value);
```

Comment:
I may be misunderstanding this. Why is a temporary needed here? Also, setLow just copies the value in `curr->value`, which is 64-bit, don't we need to truncate it to get the low bits?

Reply:
I think I see, thanks. What I believe happens is the lowering was already done, so this node is 32-bit, and the high 32 bits are received in `fetchOutParam`.

If that's right, then you don't need the `set_local`/`get_local` (or the block) here, since the low bits are used exactly once, you can just use them directly.

Reply:
I mean, I believe there is no need for the temporary variable. It could be `auto lowBits = curr->value;`

Reply:
How is that any different from the original code, which used `curr->value` directly?

Reply:
I think the original code got the part of receiving the low and high bits right, but it just did an AND operation on them, which your PR fixes, it should indeed by an EQZ of an OR.

Reply:
The original code wanted to do `(and (eqz $hi) (eqz $lo))` which is a valid translation of `i64.eqz`: if both 32-bit parts are equal to zero, the 64-bit integer is equal to zero.  But the code as written was generating the above asm.js, which was wrong.

Even with the `(eqz (or $hi $lo))` algorithm, if you change it to use a bare `curr->value`:

```c++
    TempVar highBits = fetchOutParam(curr->value);

    Block* result = builder->blockify(
      builder->makeUnary(
        EqZInt32,
        builder->makeBinary(
          OrInt32,
          builder->makeGetLocal(highBits, i32),
          curr->value
        )
      )
    );

    replaceCurrent(result);
```

You get incorrect results, because the asm.js looks like:

```js
 function $5($0, $0$hi) {
  $0 = $0 | 0;
  $0$hi = $0$hi | 0;
  var i64toi32_i32$0 = 0, $3 = 0, $4 = 0, $5 = 0, $6 = 0, $7 = 0, $8 = 0, $9 = 0, $10 = 0, $11 = 0;
  $3 = i64toi32_i32$0;
  i64toi32_i32$0 = $0$hi;
  return ($3 | $0 | 0 | 0) == (0 | 0) | 0;
 }
```

which is pretty bogus.

Reply:
I think that's due to an ordering bug, see https://github.com/WebAssembly/binaryen/commit/3572e575322024c49737a25c654437b08c531140 . The issue is that `curr->value` sets up both the low and high bits, so it needs to execute first before we use either. And in an or the order doesn't matter so we can just flip it.

Btw, I think it's easier to understand looking at the pass by itself as in that commit, and testing the pass directly, what do you think?